File: sssapril/vov/server/app/api/v1/resources.py
"""
资源API路由模块

提供资源相关的REST API端点。
"""

from typing import List, Optional
from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.ext.asyncio import AsyncSession

from app.api.deps import get_db
from app.services.resource_service import ResourceService
from app.schemas.resource import (
    ResourceCreate,
    ResourceUpdate,
    ResourceResponse,
)
from app.schemas.common import ApiResponse, PaginatedResponse, StatusResponse

router = APIRouter(tags=["resources"])


# 项目全局资源端点
@router.get("/projects/{project_id}/resources", response_model=ApiResponse[PaginatedResponse[ResourceResponse]])
async def list_project_resources(
    project_id: str,
    resource_type: Optional[str] = Query(None, description="资源类型筛选"),
    is_required: Optional[bool] = Query(None, description="是否必读筛选"),
    db: AsyncSession = Depends(get_db),
):
    """
    获取项目的全局资源列表

    返回项目中所有全局资源（非群聊专属）。
    """
    service = ResourceService(db)
    resources = await service.get_by_project(project_id, resource_type, is_required)
    return ApiResponse(data=PaginatedResponse(items=resources, total=len(resources), skip=0, limit=100))


@router.get("/projects/{project_id}/resources/required", response_model=ApiResponse[PaginatedResponse[ResourceResponse]])
async def list_required_resources(
    project_id: str,
    db: AsyncSession = Depends(get_db),
):
    """
    获取项目的必读资源

    返回项目中所有标记为必读的资源。
    """
    service = ResourceService(db)
    resources = await service.get_required_by_project(project_id)
    return ApiResponse(data=PaginatedResponse(items=resources, total=len(resources), skip=0, limit=100))


# 群聊资源端点
@router.get("/groups/{group_id}/resources", response_model=ApiResponse[PaginatedResponse[ResourceResponse]])
async def list_group_resources(
    group_id: str,
    db: AsyncSession = Depends(get_db),
):
    """
    获取群聊的资源列表

    返回群聊专属的资源。
    """
    service = ResourceService(db)
    resources = await service.get_by_group(group_id)
    return ApiResponse(data=PaginatedResponse(items=resources, total=len(resources), skip=0, limit=100))


# 资源CRUD端点
@router.post("/resources", response_model=ApiResponse[ResourceResponse], status_code=201)
async def create_resource(
    data: ResourceCreate,
    db: AsyncSession = Depends(get_db),
):
    """
    创建资源

    创建新资源。
    """
    service = ResourceService(db)
    resource = await service.create_resource(data.model_dump())
    return ApiResponse(data=resource)


@router.get("/resources/{resource_id}", response_model=ApiResponse[ResourceResponse])
async def get_resource(
    resource_id: str,
    db: AsyncSession = Depends(get_db),
):
    """
    获取资源详情

    返回资源的详细信息。
    """
    service = ResourceService(db)
    resource = await service.get_by_id(resource_id)
    if not resource:
        raise HTTPException(status_code=404, detail="Resource not found")
    return ApiResponse(data=resource)


@router.put("/resources/{resource_id}", response_model=ApiResponse[ResourceResponse])
async def update_resource(
    resource_id: str,
    data: ResourceUpdate,
    db: AsyncSession = Depends(get_db),
):
    """
    更新资源

    更新资源信息。
    """
    service = ResourceService(db)
    resource = await service.update_resource(resource_id, data.model_dump(exclude_unset=True))
    if not resource:
        raise HTTPException(status_code=404, detail="Resource not found")
    return ApiResponse(data=resource)


@router.delete("/resources/{resource_id}", response_model=ApiResponse[StatusResponse])
async def delete_resource(
    resource_id: str,
    db: AsyncSession = Depends(get_db),
):
    """
    删除资源

    删除指定资源（软删除）。
    """
    service = ResourceService(db)
    success = await service.delete_resource(resource_id)
    if not success:
        raise HTTPException(status_code=404, detail="Resource not found")
    return ApiResponse(data=StatusResponse(success=True, message="Resource deleted"))

File: sssapril/vov/client/src/api/resources.ts
/**
 * 资源API模块
 *
 * 封装资源(Resource)和标签(Tag)相关的API调用。
 */

import { apiClient } from './client';
import {
  ApiResponse,
  PaginatedResponse,
  Resource,
  CreateResourceRequest,
  UpdateResourceRequest,
  Tag,
  CreateTagRequest,
  UpdateTagRequest,
} from '../types';

/**
 * 资源API
 */
export const resourceApi = {
  /**
   * 获取项目全局资源列表
   *
   * @param projectId - 项目ID
   * @param type - 资源类型筛选
   * @param required - 是否必读筛选
   * @returns 资源列表
   */
  listByProject(projectId: string, type?: string, required?: boolean) {
    return apiClient.get<PaginatedResponse<Resource>>(`/projects/${projectId}/resources`, {
      type,
      required: required !== undefined ? String(required) : undefined,
    });
  },

  /**
   * 获取群聊资源列表
   *
   * @param groupId - 群聊ID
   * @returns 资源列表
   */
  listByGroup(groupId: string) {
    return apiClient.get<PaginatedResponse<Resource>>(`/groups/${groupId}/resources`);
  },

  /**
   * 创建资源
   *
   * @param data - 创建参数
   * @returns 创建的资源
   */
  create(data: CreateResourceRequest) {
    return apiClient.post<Resource>('/resources', data);
  },

  /**
   * 更新资源
   *
   * @param id - 资源ID
   * @param data - 更新参数
   * @returns 更新后的资源
   */
  update(id: string, data: UpdateResourceRequest) {
    return apiClient.put<Resource>(`/resources/${id}`, data);
  },

  /**
   * 删除资源
   *
   * @param id - 资源ID
   */
  delete(id: string) {
    return apiClient.delete<void>(`/resources/${id}`);
  },
};

/**
 * 标签API
 */
export const tagApi = {
  /**
   * 获取项目标签列表
   *
   * @param projectId - 项目ID
   * @returns 标签列表
   */
  list(projectId: string) {
    return apiClient.get<PaginatedResponse<Tag>>(`/projects/${projectId}/tags`);
  },

  /**
   * 创建标签
   *
   * @param projectId - 项目ID
   * @param data - 创建参数
   * @returns 创建的标签
   */
  create(projectId: string, data: CreateTagRequest) {
    return apiClient.post<Tag>(`/projects/${projectId}/tags`, data);
  },

  /**
   * 更新标签
   *
   * @param id - 标签ID
   * @param data - 更新参数
   * @returns 更新后的标签
   */
  update(id: string, data: UpdateTagRequest) {
    return apiClient.put<Tag>(`/tags/${id}`, data);
  },

  /**
   * 删除标签
   *
   * @param id - 标签ID
   */
  delete(id: string) {
    return apiClient.delete<void>(`/tags/${id}`);
  },
};

export default { resourceApi, tagApi };
